---
title: "Assignment 2: Differential Gene Expression and Preliminary ORA"
author: "Vishali Umaiyalan"
output:
  html_document:
    toc: true
    toc_depth: 6
bibliography: a2_references.bib
csl: biomed-central.csl
---

# 0. Libraries
```{r, message=FALSE}
require(edgeR)
require(ggplot2)
require(ComplexHeatmap)
require(gprofiler2)
require(knitr)
```

# 1. Introduction 
The chosen dataset comes from "Integrated functional genomic analyses of Klinefelter and Turner syndromes reveal global network effects of altered X chromosome dosage" which was published in February, 2020 by Zhang et. al. We will conduct differential expression analysis, using `edgeR`, between control individuals and individuals with either Klinefelter or Turner syndrome (KS and TS, respectively). As a reminder, KS occurs in males when they are born with an extra X chromosome (i.e. XXY), while TS occurs in females when they are born missing an X chromosome (i.e. XO). Specifically, we will identify differentially expressed genes between males with KS and control males and differentially expressed genes between females with TS and control females. Subsequently, we will perform thresholded over-representation analysis on each gene list. 

# 2. Load the normalized counts from Assignment 1
For Assignment 1, we filtered out genes there were not expressed at more than 1 count per million (CPM) in at least 3 samples, and normalized the filtered counts using TMM. Here we load in the resulting counts.

```{r, message=FALSE}
# Create a data folder in the parent directory of the a2 directory
dest_dir <- file.path(getwd(), "data")
if (! dir.exists(dest_dir)) {
  dir.create(dest_dir)
}

# Download the counts if they have not already been downloaded
counts_path <- file.path(dest_dir, "normalized_counts.rds")
if (! file.exists(counts_path)) {
  file_url <- "https://github.com/bcb420-2024/Vishali_Umaiyalan/blob/main/data/normalized_counts.rds?raw=true"
  download.file(url = file_url, destfile = counts_path)
}

# Load in the normalized counts
normalized_counts <- readRDS(file = counts_path)
```

# 3. Prepare the data for the DE analyses
```{r}
# Make a list of the sample types for each column in the counts
sample_type <- rep("control", ncol(normalized_counts))
sample_type[grep("_KS", colnames(normalized_counts))] <- "KS"
sample_type[grep("_TS", colnames(normalized_counts))] <- "TS"
sample_type[grep("control_male", colnames(normalized_counts))] <- "control_male"
sample_type[grep("control_female", colnames(normalized_counts))] <- "control_female"

# Create separate counts for males and females
male_indices <- grep("control_male|_KS", colnames(normalized_counts))
female_indices <- grep("control_female|_TS", colnames(normalized_counts))
male_counts <- normalized_counts[, male_indices]
female_counts <- normalized_counts[, female_indices]

# Make a design matrix for each analysis: 
# 1. Control males vs. males with KS
# 2. Control females vs. females with TS
# There are no replicates in this data set

# The controls will be the baseline for both analyses, so control should not
# be a column in the design matrices
male_sample_type <- sample_type[male_indices]
design_mat_male <- model.matrix(~male_sample_type)

female_sample_type <- sample_type[female_indices]
design_mat_females <- model.matrix(~female_sample_type)

# Construct the DGEList objects from the counts
dge_male <- DGEList(counts = male_counts)
dge_female <- DGEList(counts = female_counts)
```
Now we have constructed design matrices and DGEList objects for males and females. There are 13 control males, 13 control females, 14 males with KS, and 14 females with TS. 

# 4. Create MDS plot
```{r}
mds_male <- plotMDS(dge_male, label = NULL, pch = 1,
                    col = c("darkgreen", "purple")[factor(male_sample_type)],
                    main = "Fig. 1a: MDS Plot for Males")
legend("topright", legend = levels(factor(male_sample_type)), pch = c(1),
       col = c("darkgreen", "purple"), title = "Disease Group")
```
<br>
Figure 1a: This MDS plot compares control males and males with KS. There is a bit of separation between the controls and KS individuals, with most controls grouping together, and the KS individuals grouping together.

```{r}
mds_female <- plotMDS(dge_female, label = NULL, pch = 1,
                    col = c("darkgreen", "purple")[factor(female_sample_type)],
                    main = "Fig. 1b: MDS Plot for Females")
legend("topright", legend = levels(factor(female_sample_type)), pch = c(1),
       col = c("darkgreen", "purple"), title = "Disease Group")
```
<br>
Figure 1b: This MDS plot compares control females and females with TS. There is a bit of separation between the controls and TS individuals, with most controls grouping together, and the KS individuals grouping together. The separation for females is not as strong as the separation for males.

# 5. Estimate dispersion
```{r}
dge_male <- estimateDisp(dge_male, design_mat_male)
bcv_male <- plotBCV(dge_male, main = "Fig. 2a: BCV Plot for Males")
```
<br>
Figure 2a: This BCV plot gives an overview on the variability in expression of genes for males. The common dispersion (red line) is relatively low (less than 0.25). 

```{r}
dge_female <- estimateDisp(dge_female, design_mat_females)
bcv_female <- plotBCV(dge_female, main = "Fig. 2b: BCV Plot for Females")
```
<br>
Figure 2b: This BCV plot gives an overview on the variability in expression of genes for females. The common dispersion (red line) is also relatively low (less than 0.25). 

# 6. Identify differentially expressed genes
```{r}
# Fit the model
fit_male <- glmQLFit(dge_male, design_mat_male)
fit_female <- glmQLFit(dge_female, design_mat_females)

# Perform quasi-likelihood F-tests
qlf_male <- glmQLFTest(fit_male)
qlf_female <- glmQLFTest(fit_female)

# Get the top hits 
top_hits_male <- topTags(qlf_male, sort.by = "PValue", n = nrow(male_counts))
top_hits_female <- topTags(qlf_female, sort.by = "PValue", n = nrow(female_counts))
```

# 7. Interpret differential expression results
```{r}
# Identify the number of genes that are significantly differentially expressed
# before multiple testing correction
sprintf("Males: %s", length(which(top_hits_male$table$PValue < 0.05)))
sprintf("Females: %s", length(which(top_hits_female$table$PValue < 0.05)))
```
1. How many genes were significantly differentially expressed? What thresholds did you use and why? \
There were 511 genes that were significantly differentially expressed when comparing control males and males with KS. There were 1,998 genes that were significantly differentially expressed when comparing control females and females with TS. A threshold of 0.05 was used for both cases. Using 0.05 retains a reasonable number of genes. Using 0.01 may also have been reasonable, but we went with 0.05 so we would have more genes in our analysis. 

```{r}
# Identify the number of genes that are significantly differentially expressed
# after multiple testing correction
sprintf("Males: %s", length(which(top_hits_male$table$FDR < 0.05)))
sprintf("Females: %s", length(which(top_hits_female$table$FDR < 0.05)))
```
2. Which method did you use for multiple testing correction? And Why? How many genes passed correction? \
The Benjimini-Hochberg method was used for multiple testing correction because it seems widely adopted for differential expression. 19 genes in males and 198 genes in females passed testing correction.

```{r}
# Create an MA plot for males
is_de <- rownames(top_hits_male$table[abs(top_hits_male$table$logFC) > 1,])
plotSmear(qlf_male, main = "Fig. 3a: MA Plot Comparing Control Males and Males with KS",
          de.tags = is_de)
abline(h = c(-1, 1), col = "blue")
```
<br>
Figure 3a: This MA plot shows the log fold change vs. the average expression between control males and males with KS, with genes that have an absolute log fold change greater than 1 colored in red. 

```{r}
# Create an MA plot for females
is_de <- rownames(top_hits_female$table[abs(top_hits_female$table$logFC) > 1,])
plotSmear(qlf_female, main = "Fig. 3b: MA Plot Comparing Control Females and Females with TS",
          de.tags = is_de)
abline(h = c(-1, 1), col = "blue")
```
<br>
Figure 3b: This MA plot shows the log fold change vs. the average expression between control females and females with TS, with genes that have an absolute log fold change greater than 1 colored in red. Control females and females with TS have more differentially expressed genes than control males and males with KS.  

```{r}
# Now create volcano plots to visualize the genes that are significantly 
# differentially expressed 
results <- as.data.frame(top_hits_male$table)

# Create a new column in the results data frame for significance
results$Significant <- with(results, logFC > 1 & FDR < 0.05 | logFC < -1 & FDR < 0.05)

# Create the Volcano plot
v_plot_males <- ggplot(results, aes(x = logFC, y = -log10(PValue), colour = Significant)) +
  geom_point(alpha = 0.5) + 
  scale_color_manual(values = c("black", "red")) +  # Non-significant points in black, significant in red
  theme_minimal() +
  labs(title = "Fig. 4a: Differential Expression between Control Males and Males with KS",
       x = "Log Fold Change",
       y = "-Log10 P-value") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "blue") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue") 

kable(results[results$Significant == TRUE,], format = "html") # print the results
v_plot_males # print the graph 
```

Figure 4a: The volcano plot shows the genes that have a log fold change above the threshold, and an adjusted p-value that is less than threshold for control males vs. males with KS. The p-value is plotted on the y-axis and the log fold change is plotted on the x-axis. If the adjusted p-value is less than 0.05, the point is colored in red. Before testing correction, there are several genes that have a log fold change above the threshold and a p-value less than 0.05 (these points can be seen past the threshold dotted lines). However, after testing correction, only one gene passes both thresholds, and this gene is shown in red, and is also summarized in the table. The gene is XIST. 

```{r}
# Create volcano plots to visualize the genes that are significantly 
# differentially expressed (females)
results <- as.data.frame(top_hits_female$table)

# Create a new column in the results data frame for significance
results$Significant <- with(results, logFC > 1 & FDR < 0.05 | logFC < -1 & FDR < 0.05)

# Create the Volcano plot
v_plot_females <- ggplot(results, aes(x = logFC, y = -log10(PValue), colour = Significant)) +
  geom_point(alpha = 0.5) + 
  scale_color_manual(values = c("black", "red")) +  # Non-significant points in black, significant in red
  theme_minimal() +
  labs(title = "Fig. 4b: Differential Expression between Control Females and Females with TS",
       x = "Log Fold Change",
       y = "-Log10 P-value") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "blue") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue") 

kable(head(results[results$Significant == TRUE,], 20), format = "html") # print the top 20 results 
v_plot_females # print the volcano plot
```
<br>
Figure 4b: The volcano plot shows the genes that have a log fold change above the threshold, and an adjusted p-value that is less than threshold for control females vs. females with TS. Before testing correction, there are several genes that have a log fold change above the threshold and a p-value less than 0.05 (these points can be seen past the threshold dotted lines). However, after testing correction, fewer genes pass both thresholds, and those genes are shown in red. There are 40 genes that pass both thresholds, and these genes are summarized in the table. XIST is differentially expressed in both analyses.

3. Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest. \
MA plots and Volcano plots are shown in Figure 3 and Figure 4 respectively. Genes of interest are highlighted in red. 

```{r}
# Create a heatmap to visualize top hits (males)
top_genes <- head(order(top_hits_male$table$PValue), 50)
top_genes_data <- male_counts[top_genes, ]

# Scale the data and plot
top_genes_scaled <- t(scale(t(top_genes_data)))
hm_males <- Heatmap(top_genes_scaled,
                    name = "expression",
                    cluster_rows = TRUE,
                    cluster_columns = TRUE,
                    show_row_names = FALSE,
                    show_column_names = TRUE,
                    column_title = "Samples",
                    row_title = "Top Genes",
                    column_names_side = "bottom",
                    heatmap_legend_param = list(title = "Expression", at = c(-2, 0, 2), 
                                                labels = c("Low", "Medium", "High")))
draw(hm_males)
```
<br>
Figure 5a: Expression for Top 50 Differentially Expressed Genes Across Control Males and Males with KS 

```{r}
# Create a heatmap to visualize top hits (females)
top_genes <- head(order(top_hits_female$table$PValue), 50)
top_genes_data <- female_counts[top_genes, ]

# Scale the data and plot
top_genes_scaled <- t(scale(t(top_genes_data)))
hm_females <- Heatmap(top_genes_scaled,
                    name = "expression",
                    cluster_rows = TRUE,
                    cluster_columns = TRUE,
                    show_row_names = FALSE,
                    show_column_names = TRUE,
                    column_title = "Samples",
                    row_title = "Top Genes",
                    column_names_side = "bottom",
                    heatmap_legend_param = list(title = "Expression", at = c(-2, 0, 2), 
                                                labels = c("Low", "Medium", "High")))
draw(hm_females)
```
<br>
Figure 5b: Expression for Top 50 Differentially Expressed Genes Across Control Females and Females with TS 

4. Visualize your top hits using a heatmap. Do you conditions cluster together? Explain why or why not. \
Heatmaps for males and females are shown in Figure 5. In males, the conditions do not seem to cluster together. The dendrogram at the top groups together control males and males with KS randomly, and it does not seem like there is a clear clustering of controls together and males with KS together. The same is true for females. 

# 8. Perform ORA using g:profiler on all genes
```{r}
gprofiler_results_male <- gost(query = rownames(subset(top_hits_male$table, FDR < 0.05)),
                               significant=FALSE,
                               ordered_query = FALSE,
                               exclude_iea=TRUE,
                               correction_method = "fdr",
                               organism = "hsapiens",
                               source = c("REAC","WP","GO:BP"))
enrichment_results_male <- gprofiler_results_male$result

# Filter out non-significant terms for each source
significant_bp <- enrichment_results_male[enrichment_results_male$source == "GO:BP" & 
                                            enrichment_results_male$p_value < 0.05, ]
significant_reac <- enrichment_results_male[enrichment_results_male$source == "REAC" & 
                                              enrichment_results_male$p_value < 0.05, ]
significant_wp <- enrichment_results_male[enrichment_results_male$source == "WP" & 
                                            enrichment_results_male$p_value < 0.05, ]

# Get the top 5 hits for each source, or fewer if less than 5 are available after filtering
top_bp <- head(significant_bp[order(significant_bp$p_value), ], 5)
top_reac <- head(significant_reac[order(significant_reac$p_value), ], 5)
top_wp <- head(significant_wp[order(significant_wp$p_value), ], 5)

# Combine the top results and return a table with the desired columns
top_combined <- rbind(top_bp, top_reac, top_wp)
return_table <- top_combined[, c("term_name", "source", "p_value")]
rownames(return_table) <- NULL

knitr::kable(return_table, format = "html")
```
<br>
ORA Table for Control Males vs. Males with KS: g:profiler was run on a list of all significantly differentially expressed genes between control males and males wiht KS. The top 5 terms per source, along with the p-value, are shown above. 

```{r}
# Get the number of genesets per source that pass the 0.05 threshold for males
significant_terms <- subset(enrichment_results_male, p_value < 0.05)
source_counts <- table(significant_terms$source)
print(source_counts)
```

```{r}
# Repeat for females
gprofiler_results_female <- gost(query = rownames(subset(top_hits_female$table, FDR < 0.05)),
                               significant=FALSE,
                               ordered_query = FALSE,
                               exclude_iea=TRUE,
                               correction_method = "fdr",
                               organism = "hsapiens",
                               source = c("REAC","WP","GO:BP"))
enrichment_results_female <- gprofiler_results_female$result

# Filter out non-significant terms for each source
significant_bp <- enrichment_results_female[enrichment_results_female$source == "GO:BP" & 
                                            enrichment_results_female$p_value < 0.05, ]
significant_reac <- enrichment_results_female[enrichment_results_female$source == "REAC" & 
                                              enrichment_results_female$p_value < 0.05, ]
significant_wp <- enrichment_results_female[enrichment_results_female$source == "WP" & 
                                            enrichment_results_female$p_value < 0.05, ]

# Get the top 5 hits for each source, or fewer if less than 5 are available after filtering
top_bp <- head(significant_bp[order(significant_bp$p_value), ], 5)
top_reac <- head(significant_reac[order(significant_reac$p_value), ], 5)
top_wp <- head(significant_wp[order(significant_wp$p_value), ], 5)

# Combine the top results and return a table with the desired columns
top_combined <- rbind(top_bp, top_reac, top_wp)
return_table <- top_combined[, c("term_name", "source", "p_value")]
rownames(return_table) <- NULL

kable(return_table, format = "html")
```
<br>
ORA Table for Control Females vs. Females with TS: g:profiler was run on a list of all significantly differentially expressed genes between control females and females with TS The top 5 terms per source, along with the p-value, are shown above. 

```{r}
# Get the number of genesets per source that pass the 0.05 threshold for females
significant_terms <- subset(enrichment_results_female, p_value < 0.05)
source_counts <- table(significant_terms$source)
print(source_counts)
```

# 9. Perform separate ORA using g:profiler on the upregulated genes and the downregulated genes
```{r}
# For males
upregulated_genes_males <- subset(top_hits_male$table, FDR < 0.05 & logFC > 0)
downregulated_genes_males <- subset(top_hits_male$table, FDR < 0.05 & logFC < 0)

# For females
upregulated_genes_females <- subset(top_hits_female$table, FDR < 0.05 & logFC > 0)
downregulated_genes_females <- subset(top_hits_female$table, FDR < 0.05 & logFC < 0)
```

```{r}
# Start with males
gprofiler_results_male_up <- gost(query = rownames(upregulated_genes_males),
                               significant=FALSE,
                               ordered_query = FALSE,
                               exclude_iea=TRUE,
                               correction_method = "fdr",
                               organism = "hsapiens",
                               source = c("REAC","WP","GO:BP"))

# Get the results table for males
enrichment_results_male_up <- gprofiler_results_male_up$result

# Filter out non-significant terms for each source
significant_bp <- enrichment_results_male_up[enrichment_results_male_up$source == "GO:BP" & 
                                            enrichment_results_male_up$p_value < 0.05, ]
significant_reac <- enrichment_results_male_up[enrichment_results_male_up$source == "REAC" & 
                                              enrichment_results_male_up$p_value < 0.05, ]
significant_wp <- enrichment_results_male_up[enrichment_results_male_up$source == "WP" & 
                                            enrichment_results_male_up$p_value < 0.05, ]

# Get the top 5 hits for each source, or fewer if less than 5 are available after filtering
top_bp <- head(significant_bp[order(significant_bp$p_value), ], 5)
top_reac <- head(significant_reac[order(significant_reac$p_value), ], 5)
top_wp <- head(significant_wp[order(significant_wp$p_value), ], 5)

# Combine the top results and return a table with the desired columns
top_combined <- rbind(top_bp, top_reac, top_wp)
return_table <- top_combined[, c("term_name", "source", "p_value")]
rownames(return_table) <- NULL

kable(return_table, format = "html")
```
<br>
ORA Table for Upregulated Genes between Control Males vs. Males with KS: g:profiler was run on a list of all upregulated significant genes between control males and males with KS. The top 5 terms per source, along with the p-value, are shown above.

```{r}
# Now downregulated
gprofiler_results_male_down <- gost(query = rownames(downregulated_genes_males),
                               significant=FALSE,
                               ordered_query = FALSE,
                               exclude_iea=TRUE,
                               correction_method = "fdr",
                               organism = "hsapiens",
                               source = c("REAC","WP","GO:BP"))

# Get the results table for males
enrichment_results_male_down <- gprofiler_results_male_down$result

# Filter out non-significant terms for each source
significant_bp <- enrichment_results_male_down[enrichment_results_male_down$source == "GO:BP" & 
                                            enrichment_results_male_down$p_value < 0.05, ]
significant_reac <- enrichment_results_male_down[enrichment_results_male_down$source == "REAC" & 
                                              enrichment_results_male_down$p_value < 0.05, ]
significant_wp <- enrichment_results_male_down[enrichment_results_male_down$source == "WP" & 
                                            enrichment_results_male_down$p_value < 0.05, ]

# Get the top 5 hits for each source, or fewer if less than 5 are available after filtering
top_bp <- head(significant_bp[order(significant_bp$p_value), ], 5)
top_reac <- head(significant_reac[order(significant_reac$p_value), ], 5)
top_wp <- head(significant_wp[order(significant_wp$p_value), ], 5)

# Combine the top results and return a table with the desired columns
top_combined <- rbind(top_bp, top_reac, top_wp)
return_table <- top_combined[, c("term_name", "source", "p_value")]
rownames(return_table) <- NULL

kable(return_table, format = "html")
```
<br>
ORA Table for Downregulated Genes between Control Males vs. Males with KS: g:profiler was run on a list of all downregulated significant genes between control males and males with KS. The top 5 terms per source, along with the p-value, are shown above.

```{r}
# Repeat for females
gprofiler_results_female_up <- gost(query = rownames(upregulated_genes_females),
                               significant=FALSE,
                               ordered_query = FALSE,
                               exclude_iea=TRUE,
                               correction_method = "fdr",
                               organism = "hsapiens",
                               source = c("REAC","WP","GO:BP"))

# Get the results table for males
enrichment_results_female_up <- gprofiler_results_female_up$result

# Filter out non-significant terms for each source
significant_bp <- enrichment_results_female_up[enrichment_results_female_up$source == "GO:BP" & 
                                            enrichment_results_female_up$p_value < 0.05, ]
significant_reac <- enrichment_results_female_up[enrichment_results_female_up$source == "REAC" & 
                                              enrichment_results_female_up$p_value < 0.05, ]
significant_wp <- enrichment_results_female_up[enrichment_results_female_up$source == "WP" & 
                                            enrichment_results_female_up$p_value < 0.05, ]

# Get the top 5 hits for each source, or fewer if less than 5 are available after filtering
top_bp <- head(significant_bp[order(significant_bp$p_value), ], 5)
top_reac <- head(significant_reac[order(significant_reac$p_value), ], 5)
top_wp <- head(significant_wp[order(significant_wp$p_value), ], 5)

# Combine the top results and return a table with the desired columns
top_combined <- rbind(top_bp, top_reac, top_wp)
return_table <- top_combined[, c("term_name", "source", "p_value")]
rownames(return_table) <- NULL

kable(return_table, format = "html")
```
<br>
ORA Table for Upregulated Genes between Control Females vs. Females with TS: g:profiler was run on a list of all upregulated significant genes between control females and females with TS. The top 5 terms per source, along with the p-value, are shown above.

```{r}
# Now downregulated
gprofiler_results_female_down <- gost(query = rownames(downregulated_genes_females),
                               significant=FALSE,
                               ordered_query = FALSE,
                               exclude_iea=TRUE,
                               correction_method = "fdr",
                               organism = "hsapiens",
                               source = c("REAC","WP","GO:BP"))

# Get the results table for males
enrichment_results_female_down <- gprofiler_results_female_down$result

# Filter out non-significant terms for each source
significant_bp <- enrichment_results_female_down[enrichment_results_female_down$source == "GO:BP" & 
                                            enrichment_results_female_down$p_value < 0.05, ]
significant_reac <- enrichment_results_female_down[enrichment_results_female_down$source == "REAC" & 
                                              enrichment_results_female_down$p_value < 0.05, ]
significant_wp <- enrichment_results_female_down[enrichment_results_female_down$source == "WP" & 
                                            enrichment_results_female_down$p_value < 0.05, ]

# Get the top 5 hits for each source, or fewer if less than 5 are available after filtering
top_bp <- head(significant_bp[order(significant_bp$p_value), ], 5)
top_reac <- head(significant_reac[order(significant_reac$p_value), ], 5)
top_wp <- head(significant_wp[order(significant_wp$p_value), ], 5)

# Combine the top results and return a table with the desired columns
top_combined <- rbind(top_bp, top_reac, top_wp)
return_table <- top_combined[, c("term_name", "source", "p_value")]
rownames(return_table) <- NULL

kable(return_table, format = "html")
```
<br>
ORA Table for Downregulated Genes between Control Females vs. Females with TS: g:profiler was run on a list of all downregulated significant genes between control females and females with TS. The top 5 terms per source, along with the p-value, are shown above.

# 10. Interpret ORA results

1. Which method did you choose and why? \
I chose g:profiler because there is an r package that can be used to run g:profiler.

2. What annotation data did you use and why? What version of the annotation are you using? \
I chose to use the same annotation sources as the ones used in the previous assignent: GO:BP, REAC, and WP. The version for all three sources is January, 2024. This was found using `get_version_info(organism='hsapiens')`. 

3. How many genesets were returned with what thresholds? \
The number of genesets per source that passed a 0.05 threshold are summarized in the two tables above. 

4. Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated and down regulated differentially expressed genes separately)? \
The analysis is shown above. These results differ significantly compared to the analysis that was run on the whole list. Importantly, the results for downregulated genes are significantly different because there are no terms from GO:BP for males and females. 

# 11. Interpretation
1. Do the over-representation results support conclusions or mechanism discussed in the original paper? \
The over-representation results do support conclusions in the original paper (@data) to some extent. For example, the paper mentions that some of the differentially expressed genes are involved in regulation. Similarly, some of the terms from the over-representation analysis includes translation initiation.

2. Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your results? \
There are significantly more upregulated genes than downregulated genes in both analyses, which supports the results from the original paper (@data). Another publication (@background) suggests that most of the differentially expressed genes between controls and females with TS are involved with hematologic/immune system or metabolism. This could supports the results in the ORA Table for Upregulated Genes between Control Females vs. Females with TS, since some of the terms relate to the hematologic system.

# 12. References
@edgeR @ggplot2 @ComplexHeatmap @gprofiler @knitr @data @background
